src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -191,6 +194,47 @@ def _extract_accessed_domains(
     return accessed_domains
 
 
+def _restrict_domain(
```

Comment:
The interface, naming, contents of the following 3 functions on first sight look like they could be improved.